### ISSUE TITLE
Make PostExpectation generic.

### ISSUE THREAD

**Issue by grouma** [dart-lang/mockito#98]
An error cropping up while migrating tests to Dart 2.0 semantics is having invalid return types for `thenAnswer` which cause runtime issues.
For example consider the class Foo:

```
class Foo {
    Future<B> doSomething(A a);
}
```

Set up the mock for Foo as follows:

```
MockFoo foo = new MockFoo()
when(foo.doSomething(any(a)).thenAnswer((_) => new B())
```

The above will currently run without issue in the VM. However, the return type of `thenAnswer` is invalid as it should be returning `Future<B>` instead of the raw `B` value.  I believe we can catch these errors through analysis. `when` uses `PostExpectation` which can be made generic. This will allow us to enforce the return type of `thenAnswer` accordingly.

**Comment by srawlins** [comment]
Sounds good. Maybe specifically Answering needs to be type annotated? https://github.com/dart-lang/mockito/blob/master/lib/src/mock.dart#L605

**Comment by grouma** [comment]
I'm not sure if that will help. We need to have a return type here: https://github.com/dart-lang/mockito/blob/master/lib/src/mock.dart#L320

**Comment by srawlins** [comment]
Ah I see. Do you see any way to type the T in `PostExpectation<T>`?

**Comment by grouma** [comment]
I think the type should be coming from here: https://github.com/dart-lang/mockito/blob/master/lib/src/mock.dart#L744

**Comment by srawlins** [comment]
Yeah I can see it. Sounds good, I'll accept a PR.

**Comment by srawlins** [comment]
Fixed with #101
